packages/core/src/VDocument.ts

```diff
@@ -152,4 +153,199 @@ export class VDocument {
+     *
+     * @param type
+     */
+    toggleList(type: ListType): void {
+        // Retrieve the nodes in the selection.
+        let selectedNodes: Array<VNode>;
+        if (this.range.isCollapsed()) {
+            // Toggle the parent if the range is collapsed.
+            selectedNodes = [this.range.start.parent];
+        } else {
+            selectedNodes = this.range.selectedNodes;
+        }
+
+        // Check if all selected nodes are within a list of the same type.
+        const htmlTag = type === ListType.ORDERED ? 'OL' : 'UL';
+        const areAllInMatchingList = selectedNodes.every(node => {
+            const listAncestor = this._listAncestor(node);
+            return listAncestor && listAncestor.htmlTag === htmlTag;
+        });
+
+        // Dispatch.
+        if (areAllInMatchingList) {
+            // If all selected nodes are within a list of the same type,
+            // "unlist" them.
+            this._unlist(selectedNodes);
+        } else {
+            this._convertToList(selectedNodes, htmlTag);
+        }
+    }
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Convert the given nodes to a list with the given `htmlTag`.
+     *
+     * @param nodes
+     * @param htmlTag
+     */
+    _convertToList(nodes: Array<VNode>, htmlTag: string): void {
+        // Lists cannot have atomic nodes as direct children.
+        nodes = Array.from(new Set(nodes.map(node => (node.atomic ? node.parent : node))));
+
+        // If the first and last nodes are within lists, we need to split those
+        // lists.
+        const duplicatedNodes = [];
+        const first = nodes[0];
+        const firstList = this._listAncestor(first);
+        if (firstList) {
+            let child = first;
+            first.ancestor(ancestor => {
+                duplicatedNodes.push(ancestor.splitAt(child));
+                child = ancestor;
+                return ancestor === firstList;
+            });
+            if (!firstList.hasChildren()) {
+                firstList.remove();
+            }
+        }
+        const last = nodes[nodes.length - 1];
+        const lastList = this._listAncestor(last);
+        if (lastList) {
+            let child = last;
+            last.ancestor(ancestor => {
+                duplicatedNodes.push(ancestor.splitAt(child));
+                child = ancestor;
+                return ancestor === lastList;
+            });
+            if (!lastList.hasChildren()) {
+                lastList.remove();
+            }
+        }
+
+        // Create and insert the list before the common ancestor to the first
+        // last node, unless the first node is within a list, in which case
+        // insert before the list instead to avoid inserting the new list within
+        // a list.
+        let list = new ListNode(htmlTag);
```

Comment:
This is where I need `htmlTag`. Although granted it might make more sense if `ListNode` would take a `ListType` instead...